cnf-certification-test/platform/suite.go

```diff
@@ -207,100 +206,95 @@ func testContainersFsDiff(env *provider.TestEnvironment) {
+			nodeName, taintsMask, nodetainted.DecodeKernelTaintsFromBitMask(taintsMask))
 
-			// Retrieve all of the out of tree modules.
-			taintedModules := tf.GetOutOfTreeModules(modules)
-			logrus.Debug("Collected all of the tainted modules: ", taintedModules)
-			logrus.Debug("Modules allowed via configuration: ", env.Config.AcceptedKernelTaints)
+		// Check the white list. If empty, mark this node as failed.
+		if len(whiteListedModules) == 0 {
+			errNodes = append(errNodes, nodeName)
+			continue
+		}
 
-			// Looks through the accepted taints listed in the tnf-config file.
-			// If all of the tainted modules show up in the configuration file, do not fail the test.
-			nodeTaintsAccepted = nodetainted.TaintsAccepted(env.Config.AcceptedKernelTaints, taintedModules)
+		// White list check.
+		// Get the list of modules (tainters) that have set a taint bit.
+		//   1. Each module should appear in the white list.
+		//   2. All kernel taint bits (one bit <-> one letter) should have been set by at least
+		//      one tainter module.
+		tainters, taintBitsByAllModules, err := tf.GetTainterModules(whiteListedModules)
+		if err != nil {
+			tnf.ClaimFilePrintf("failed to get tainter modules from node %s: %v", nodeName, err)
+			errNodes = append(errNodes, nodeName)
+			continue
 		}
 
-		// Only add the tainted node to the slice if the taint is acceptable.
-		if !nodeTaintsAccepted {
-			taintedNodes = append(taintedNodes, dp.Name)
+		// Save modules' names only.
+		for moduleName, taintsLetters := range tainters {
+			moduleTaints := nodetainted.DecodeKernelTaintsFromLetters(taintsLetters)
+			badModules[nodeName] = append(badModules[nodeName], badModuleTaints{name: moduleName, taints: moduleTaints})
+
+			tnf.ClaimFilePrintf("Node %s - module %s taints kernel: %s", nodeName, moduleName, moduleTaints)
 		}
 
-		// Only print the message if there is something to report failure or tainted node wise.
-		if len(taintedNodes) != 0 || len(errNodes) != 0 {
-			tnf.ClaimFilePrintf("Decoded tainted kernel causes (code=%d) for node %s : %s\n", taintedBitmap, dp.Name, strings.Join(taints, ","))
+		// Lastly, check that all kernel taint bits come from modules.
+		otherKernelTaints := nodetainted.GetOtherTaintedBits(taintsMask, taintBitsByAllModules)
+		for _, taintedBit := range otherKernelTaints {
+			tnf.ClaimFilePrintf("Node %s - taint bit %d is set but it's not caused by any module.", nodeName, taintedBit)
+			otherTaints[nodeName] = append(otherTaints[nodeName], taintedBit)
 		}
 	}
 
-	// We are expecting tainted nodes to be Nil, but only if:
-	// 1) The reason for the tainted node is contains(`module was loaded`)
-	// 2) The modules loaded are all whitelisted.
-	if len(taintedNodes) > 0 {
-		tnf.ClaimFilePrintf("Nodes have been found to be tainted: %v", taintedNodes)
-		ginkgo.Fail("Nodes have been found to be tainted.")
-	}
+	logrus.Infof("Nodes with errors: %+v", errNodes)
```

Comment:
Why not use AddTestResultLog here:
testhelper.AddTestResultLog("Nodes with errors", errNodes, tnf.ClaimFilePrintf, ginkgo.Fail)
testhelper.AddTestResultLog("Bad Modules", badModules, tnf.ClaimFilePrintf, ginkgo.Fail)
testhelper.AddTestResultLog("Taints not related to any module", otherTaints, tnf.ClaimFilePrintf, ginkgo.Fail)

Reply:
I wanted to show all the maps before failing the TC. If the first testhelper.AddTestResultLog() in your sample code fails for errNodes because of a failure with one node, the badModules/otherTaints from the other modules won't be shown in the log, as the ginkgo.Fail() called inside will stop the abort the execution of the code.

Reply:
looks good, we can add a similar feature later